source/client.c

```diff
@@ -77,6 +77,38 @@ void aws_mqtt_client_clean_up(struct aws_mqtt_client *client) {
     AWS_ZERO_STRUCT(*client);
 }
 
+/* Reconnect helpers */
+static void s_attempt_reconnect(struct aws_task *task, void *userdata, enum aws_task_status status);
+
+int s_create_reconnect_task(struct aws_mqtt_client_connection *connection) {
+    AWS_ASSERT(!connection->reconnect_task);
+    connection->reconnect_task = aws_mem_calloc(connection->allocator, 1, sizeof(struct aws_mqtt_reconnect_task));
+    if (!connection->reconnect_task) {
+        AWS_LOGF_ERROR(AWS_LS_MQTT_CLIENT, "id=%p: Failed to allocate reconnect task", (void *)connection);
+        return AWS_OP_ERR;
+    }
+    aws_atomic_init_ptr(&connection->reconnect_task->connection_ptr, connection);
+    connection->reconnect_task->allocator = connection->allocator;
+    aws_task_init(&connection->reconnect_task->task, s_attempt_reconnect, connection->reconnect_task, "mqtt_reconnect");
+
+    return AWS_OP_SUCCESS;
+}
+
+static void s_destroy_reconnect_task(struct aws_mqtt_client_connection *connection) {
+    if (connection->reconnect_task) {
+        aws_atomic_store_ptr(&connection->reconnect_task->connection_ptr, NULL);
+
+        /* Cancel if scheduled */
+        if (connection->reconnect_task->task.timestamp > 0) {
+            aws_event_loop_cancel_task(
+                aws_channel_get_event_loop(connection->slot->channel), &connection->reconnect_task->task);
```

Comment:
Is there always a channel here? Also why do you need the channel? you have the event loop group, you could even always use the same event loop for reconnect task scheduling.

Reply:
I can't find a cancel API anywhere but on the event loop or the task scheduler.

Reply:
True, the slot may have been unlinked by the time this task runs.  May need to move the destroy call into the channel shutdown callback instead.  Unsure what you mean by the second part; I don't see a way to cancel a scheduled task by any way other than the event loop it is scheduled on or the raw task task scheduler.